pkg/steps/artifacts.go

```diff
@@ -51,6 +52,72 @@ func (nopNotifier) Complete(_ string)               {}
 func (nopNotifier) Done(_ string) bool              { return true }
 func (nopNotifier) Cancel()                         {}
 
+// TestCaseNotifier allows a caller to generate per container unit tests
+// that provide better granularity for debugging problems. It intercepts
+// notifications and remembers the last pod retrieved.
+type TestCaseNotifier struct {
+	nested  ContainerNotifier
+	lastPod *coreapi.Pod
+}
+
+func NewTestCaseNotifier(nested ContainerNotifier) *TestCaseNotifier {
+	return &TestCaseNotifier{nested: nested}
+}
+
+func (n *TestCaseNotifier) Notify(pod *coreapi.Pod, containerName string) {
+	n.nested.Notify(pod, containerName)
+	n.lastPod = pod
+}
+
+func (n *TestCaseNotifier) Complete(podName string)  { n.nested.Complete(podName) }
+func (n *TestCaseNotifier) Done(podName string) bool { return n.nested.Done(podName) }
+func (n *TestCaseNotifier) Cancel()                  { n.nested.Cancel() }
+
+// SubTests returns one junit test for each terminated container with a name
+// in the annotation 'ci-operator.openshift.io/container-sub-tests' in the pod.
+// Invoking SubTests clears the last pod, so subsequent calls will return no
+// tests unless Notify() has been called in the meantime.
```

Comment:
Not sure I fully understand the point of this caching mechanism

Reply:
We use the same notifier for multiple pods in a for loop in template processing.  So if you have template with pods A and B, we pass the same notifier to A that wee do to B.  That's also how artifacts are gathered right now (we wait until A passes, then check B).

Reply:
I see